### PR TITLE
Updated istanbul consensus engine to verify headers synchronously

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- **UPDATE**: The addresses cache is updated. Previously, the `ecrecover` method used an ARU cache with block hash as a key, but we need to cache all the addresses recovered from committed seals of a block. So, 65-bytes signatures are used as keys. Furthermore, the size of the cache is increased to 2048 * 30 (max number of downloaded blocks * approximate sub group size of validators).

- After the PR https://github.com/klaytn/klaytn/pull/920, the `Refresh` method must be called all the time to refresh proposer candidates as well as validators.
- The `Refresh` method wasn't always called before because the header verification concurrently occurs while downloading a batch of blocks. (If the header does not exist, it ignores to refresh)
- The `Refresh` method didn't work properly since it was failed to get staking information to verify headers. (It tries to call a contract method on a specific block which is not yet inserted)
- This PR updates the header verification synchronously, while the computation of addresses will be processed concurrently to prevent download performance.

- The simple synchronous verification performs poorly as below panel, so extracting the pub key from seals handled concurrently.
image

- The comparision sync test is attached.
EN0 : master branch (Klaytn v1.6.1+f695312c38)
EN1 : After this PR (Klaytn v1.6.1+d76876935e)
EN2 : Before this PR (Klaytn v1.6.1+37b05fe791)

screencapture-54-180-156-230-3000-d-KnJhY-iWz-klaytn-main-overview-2021-05-17-09_20_39

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/clique/clique.go [hunk 1]
```diff
@@ -240,6 +240,16 @@ func (c *Clique) Author(header *types.Header) (common.Address, error) {
 	return ecrecover(header, c.signatures)
 }
 
+// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+func (c *Clique) CanVerifyHeadersConcurrently() bool {
+	return true
+}
+
+// PreprocessHeaderVerification prepare header verification for heavy computation before synchronous header verification such as ecrecover.
```
Comment (ehnuje) [inline consensus/clique/clique.go:248]:
```suggestion
// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
```

File: consensus/consensus.go [hunk 2]
```diff
@@ -63,6 +63,12 @@ type Engine interface {
 	// block.
 	Author(header *types.Header) (common.Address, error)
 
+	// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+	CanVerifyHeadersConcurrently() bool
+
+	// PreprocessHeaderVerification prepare header verification for heavy computation before synchronous header verification such as ecrecover.
```
Comment (ehnuje) [inline consensus/consensus.go:69]:
```suggestion
	// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
```

File: consensus/istanbul/backend/engine.go [hunk 3]
```diff
@@ -95,15 +96,68 @@ var (
...
 func (sb *backend) Author(header *types.Header) (common.Address, error) {
 	return ecrecover(header)
 }
 
+// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+func (sb *backend) CanVerifyHeadersConcurrently() bool {
+	return false
+}
+
+// PreprocessHeaderVerification prepare header verification for heavy computation before synchronous header verification such as ecrecover.
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine.go:115]:
```suggestion
// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
```

File: consensus/istanbul/backend/engine.go [hunk 4]
```diff
@@ -95,15 +96,68 @@ var (
...
+
+	proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
+	for _, seal := range istanbulExtra.CommittedSeal {
+		sealStr := hex.EncodeToString(seal)
+		if _, ok := signatureAddresses.Get(sealStr); ok {
+			continue
+		}
+		addr, err := istanbul.GetSignatureAddress(proposalSeal, seal)
+		if err != nil {
+			return err
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine.go:154]: How about returning `errInvalidSignature` in accordance with line 311?

**Review by ehnuje — COMMENTED** [review]
LGTM except for minor comments.

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -1588,7 +1588,15 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
...
-	abort, results := bc.engine.VerifyHeaders(bc, headers, seals)
+	var (
+		abort   chan<- struct{}
+		results <-chan error
+	)
+	if bc.engine.CanVerifyHeadersConcurrently() {
+		abort, results = bc.engine.VerifyHeaders(bc, headers, seals)
+	} else {
+		abort, results = bc.engine.PreprocessHeaderVerification(headers)
+	}
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:1599]: `PreprocessHeaderVerification` executes some parts of `VerifyHeaders` and the rest parts of the verfication are not executed if `bc.engine.CanVerifyHeadersConcurrently()` is false.

Comment (jeongkyun-oh) [reply blockchain/blockchain.go:1599]: Right. `VerifyHeaders` is enough to implement the functionality, but the new method is introduced since the `PreprocessHeaderVerification` does not verify headers. Any suggestion?

File: consensus/clique/clique.go (hunk 1 above)
Comment (jeongkyun-oh) [reply consensus/clique/clique.go]: Updated the comment. Thanks.

File: consensus/consensus.go (hunk 2 above)
Comment (jeongkyun-oh) [reply consensus/consensus.go]: Updated the comment. Thanks.

File: consensus/istanbul/backend/engine.go (hunk 3 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go]: Updated the comment. Thanks.

File: consensus/istanbul/backend/engine.go (hunk 4 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go]: Reflected. Thanks :)

File: consensus/istanbul/backend/engine.go [hunk 6]
```diff
@@ -95,15 +96,68 @@ var (
...
 
+// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+func (sb *backend) CanVerifyHeadersConcurrently() bool {
+	return false
+}
+
+// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
+func (sb *backend) PreprocessHeaderVerification(headers []*types.Header) (chan<- struct{}, <-chan error) {
+	abort := make(chan struct{})
+	results := make(chan error, inmemoryBlocks)
```
Comment (aidan-kwon) [inline consensus/istanbul/backend/engine.go:118]: Why did you use `inmemoryBlocks` instead of `len(headers)`?

Comment (aidan-kwon) [reply consensus/istanbul/backend/engine.go:118]: Is 100 * 30 enough cache size? Did downloader send thousands of blocks at one time?

File: consensus/istanbul/backend/engine.go [hunk 7]
```diff
@@ -95,15 +96,68 @@ var (
 	nonceAuthVote = hexutil.MustDecode("0xffffffffffffffff") // Magic nonce number to vote on adding a new validator
 	nonceDropVote = hexutil.MustDecode("0x0000000000000000") // Magic nonce number to vote on removing a validator.
 
-	inmemoryAddresses  = 20 // Number of recent addresses from ecrecover
-	recentAddresses, _ = lru.NewARC(inmemoryAddresses)
+	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
+	inmemoryValidatorsPerBlock = 30  // Approximate number of validators' addresses from ecrecover
+	recentAddresses, _         = lru.NewARC(inmemoryBlocks)
+	signatureAddresses, _      = lru.New(inmemoryBlocks * inmemoryValidatorsPerBlock)
```
Comment (aidan-kwon) [inline consensus/istanbul/backend/engine.go:102]: How about combing them into one cache and set/get items insides of `GetSignatureAddress`?

File: consensus/istanbul/backend/engine.go (hunk 6 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:118]: > Why did you use inmemoryBlocks instead of len(headers)?

This is for throttling purpose. If the computing is way faster and it replaces the already computed ones from cache, this precomputing is not useful. It may use more memory, but I think it is not necessary.

> Is 100 * 30 enough cache size? Did downloader send thousands of blocks at one time?

So, the throttling header verification every 100 blocks, and each block may contain 30 signatures at most. So, I put that number.

File: consensus/istanbul/backend/engine.go [hunk 8]
```diff
@@ -95,15 +96,68 @@ var (
...
+	// Retrieve the signature from the header extra-data
+	istanbulExtra, err := types.ExtractIstanbulExtra(header)
+	if err != nil {
+		return err
+	}
+
+	proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
+	for _, seal := range istanbulExtra.CommittedSeal {
+		sealStr := hex.EncodeToString(seal)
+		if _, ok := signatureAddresses.Get(sealStr); ok {
```
Comment (KimKyungup) [inline consensus/istanbul/backend/engine.go:149]: How about using seal as a key for `signatureAddresses`?
```suggestion
		if _, ok := signatureAddresses.Get(seal); ok {
```

File: consensus/istanbul/backend/engine.go [hunk 9]
```diff
@@ -95,15 +96,68 @@ var (
...
+			continue
+		}
+		addr, err := istanbul.GetSignatureAddress(proposalSeal, seal)
+		if err != nil {
+			return errInvalidSignature
+		}
+		signatureAddresses.Add(sealStr, addr)
```
Comment (KimKyungup) [inline consensus/istanbul/backend/engine.go:156]:
```suggestion
		addr, err := istanbul.GetSignatureAddress(proposalSeal, seal)
		if err != nil {
			return errInvalidSignature
		}
		signatureAddresses.Add(seal, addr)
```

File: consensus/istanbul/backend/engine.go [hunk 10]
```diff
@@ -95,15 +96,68 @@ var (
 	nonceAuthVote = hexutil.MustDecode("0xffffffffffffffff") // Magic nonce number to vote on adding a new validator
 	nonceDropVote = hexutil.MustDecode("0x0000000000000000") // Magic nonce number to vote on removing a validator.
 
-	inmemoryAddresses  = 20 // Number of recent addresses from ecrecover
-	recentAddresses, _ = lru.NewARC(inmemoryAddresses)
+	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
```
Comment (KimKyungup) [inline consensus/istanbul/backend/engine.go:99]: This value is enough? 
Could you check the all block header can be preprocessed in normal cases?

```suggestion
	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
```

File: consensus/istanbul/backend/engine.go (hunk 6 above)
Comment (aidan-kwon) [reply consensus/istanbul/backend/engine.go:118]: I mean throttling 100 blocks doesn't seem to be enough. Downloader could send 2048 blocks at once, but it only preprocesses 100 blocks.

Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:118]: @aidan-kwon I will increase the number to 2048. I thought the cache is not meaningful after syncing, so the space would be wasted (although only about 4MB is used).

File: consensus/istanbul/backend/engine.go (hunk 8 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:149]: @KimKyungup The byte slice is not hashable, so the raw seal cannot be used. Any other suggestion?

File: consensus/istanbul/backend/engine.go (hunk 9 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:156]: Same reason with the above comment.

File: consensus/istanbul/backend/engine.go (hunk 10 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:99]: I will increase size to 2048 suggested in https://github.com/klaytn/klaytn/pull/957/files#r629000344

File: consensus/istanbul/backend/engine.go (hunk 7 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go:102]: I think combining is fine, but the `GetSignatureAddress` is an util function, so putting the cache inside is not good. What do you think?

Comment (aidan-kwon) [reply consensus/istanbul/backend/engine.go:102]: It makes sense. You might create a wrapper function such as `CacheSignatureAddress` or put the caching logic outside as it was.

Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine.go]: I have added `cacheSignatureAddress` for caching. PTAL.

File: consensus/istanbul/backend/engine.go (hunk 8 above)
Comment (KimKyungup) [reply consensus/istanbul/backend/engine.go]: @jeongkyun-oh 
It seems that the key can be byte slice like below? Isn't it?
`hash` is byte array.
`recentAddresses.Add(hash, addr)`

File: consensus/istanbul/backend/engine.go [hunk 11]
```diff
@@ -673,22 +735,16 @@ func sigHash(header *types.Header) (hash common.Hash) {
 
 // ecrecover extracts the Klaytn account address from a signed header.
 func ecrecover(header *types.Header) (common.Address, error) {
-	hash := header.Hash()
```
Comment (KimKyungup) [inline consensus/istanbul/backend/engine.go:676]: @jeongkyun-oh Why did you remove this code?

File: consensus/istanbul/backend/engine.go (hunk 6 above)
Comment (aidan-kwon) [reply consensus/istanbul/backend/engine.go:131]: Thanks. I think `len(headers)` is more suitable for the length of `results` channel, though. `inmemoryBlocks` looks fine also.

**Comment by jeongkyun-oh** [comment]
@ehnuje @winnie-byun @yoomee1313 Please take a look.

**Review by yoomee1313 — APPROVED** [review]
lgtm

**Review by winnie-byun — APPROVED** [review]
It would be nice to put explanation. Nice work:)
EN0 : master branch (Klaytn v1.6.1+f695312c38)
EN1 : After this PR (Klaytn v1.6.1+d76876935e)
EN2 : Before this PR (Klaytn v1.6.1+37b05fe791)

**Review by ehnuje — APPROVED** [review]
